### PR TITLE
User-autentication

### LINKED ISSUE #47: Authentication Implementation
using non-native auth services to authenticate users to the application.

The authentication service makes sure that information is accessed only by users who have the access permission or right. The SpaceYaTech OpenSource CMS implements two forms of Authentication - JWT Authentication and Oauth 2.0 for Authentication and Authorization of Users within the Application. Here is a detailed look into the potential implementation of both forms of Authentication forms: 
Read more here Authentication services

### PR DESCRIPTION
**Pull request by teslima0**
# Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #47 

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: CMS/settings.py [hunk 1]
```diff
@@ -106,6 +112,56 @@
...
+    'DEFAULT_AUTHENTICATION_CLASSES': (
+        #'rest_framework_jwt.authentication.JSONWebTokenAuthentication',
+
+        'rest_framework_simplejwt.authentication.JWTAuthentication',
+     #'rest_framework.authentication.SessionAuthentication',
+       # 'rest_framework.authentication.BasicAuthentication',
+    ),
+
+    'DEFAULT_PERMISSION_CLASSES':[
+        'rest_framework.permissions.IsAuthenticated',
```
Comment (JimmyTron) [inline CMS/settings.py:144]: Why do we have this Isautheñticated  as the default value🤔🤔

File: accounts/admin.py [hunk 2]
```diff
@@ -13,11 +13,12 @@ class CustomUserAdmin(UserAdmin):
     list_display = ('username', 'email', 'is_staff', 'is_active',)
     fieldsets = (
         (None, {'fields': ('username', 'password')}),
-        ('Personal info', {'fields': ('email', 'first_name', 'last_name', 'phone_number')}),
```
Comment (JimmyTron) [inline accounts/admin.py:16]: Hope you have the phone number field back on the admin page. Haven't seen it.

**Review by JimmyTron — COMMENTED** [review]
These two have been pointed out plz checkthem  and resolve where possible

File: accounts/admin.py (hunk 2 above)
Comment (teslima0) [reply accounts/admin.py:16]: I have done that now

File: CMS/settings.py (hunk 1 above)
Comment (teslima0) [reply CMS/settings.py:144]: If Isautheñticated simply means user must login to edit their own profile.. If we don't set it, it means no login is required for anybody to make a change on our system. It means I can possibly edit other people's profile without their permission which not what we intended to achieve in my opinion

**Review by JimmyTron — CHANGES_REQUESTED** [review]
Update the requirements file to include the libraries you are using

File: CMS/settings.py [hunk 3]
```diff
@@ -106,6 +112,56 @@
...
+
+    'DEFAULT_PERMISSION_CLASSES':[
+        'rest_framework.permissions.IsAuthenticated',                 
+    ],
+    
+   'NON_FIELD_ERRORS_KEY': 'error',
+    
+}
+SIMPLE_JWT = {
+    'ACCESS_TOKEN_LIFETIME': timedelta(minutes=10),
```
Comment (tallninja) [inline CMS/settings.py:151]: Could you lower the access token lifetime and try to keep it under a minute. The reason being ten minutes is long enough for a malicious attacker to hijack the token and cause some damage.

File: accounts/serializer.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    
+    class Meta:
+        model=User
+        fields=['email','password','username','tokens']
+
+    def validate(self, attrs):
+        email = attrs.get('email', '')
+        password=attrs.get('password', '')
+
+        user = auth.authenticate(email=email, password=password)
```
Comment (tallninja) [inline accounts/serializer.py:92]: I see this is where the user gets authenticated and you are using **rest_framework_simplejwt** to do all the heavy lifting for you. From the docs, is **rest_framework_simplejwt** storing the refresh token as a [**http_only**](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies#security) cookie ? Or you just send the access and refresh tokens as a response object directly to the user ?

**Review by tallninja — COMMENTED** [review]
Everything seems fine I don't know how **rest_framework_simplejwt** works but it seems to be doing all the heavy lifting which is fine.

File: CMS/urls.py [hunk 5]
```diff
@@ -14,15 +14,46 @@
...
+   ),
+   public=True,
+   permission_classes=[permissions.AllowAny],
+)
+
 urlpatterns = [
     path('admin/', admin.site.urls),
-] + static(settings.MEDIA_URL,
-document_root=settings.MEDIA_ROOT)
+    path('ApiAuth/', include(router.urls)),
```
Comment (tallninja) [inline CMS/urls.py:52]:
```suggestion
    path('api-auth/', include(router.urls)),
```
**OR**

```suggestion
    path('auth/', include(router.urls)),
```

Conventional way of naming API Endpoints. Camel-casing is not advisable due to readability. You can find detailed guidelines [here](https://nordicapis.com/10-best-practices-for-naming-api-endpoints/)

File: CMS/urls.py [hunk 6]
```diff
@@ -14,15 +14,46 @@
...
+    path('accounts/', include('accounts.urls')),
+
+    path ('', schema_view.with_ui('swagger', cache_timeout=0), name='schema-swagger-ui'),
+    path ('api/api.json', schema_view.without_ui( cache_timeout=0), name='schema-swagger-ui'),
+    path ('redoc/', schema_view.with_ui('redoc', cache_timeout=0), name='schema-redoc'),
+    
+] + static(settings.MEDIA_URL,document_root=settings.MEDIA_ROOT)
```
Comment (tallninja) [inline CMS/urls.py:59]: Is there a **refresh-token** endpoint ? Where a user can get a new access token after it expires ?

**Review by tallninja — COMMENTED** [review]
Just some minor comments I left out in my previous review

**Review by Collins-Omariba — COMMENTED** [review]
Why is there a large block of commented out code in the serializer.py file under accounts ?

File: blog/models.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+
+from django.db import models
+from django.contrib.auth.models import User
+
+class Account(models.Model):
+    user = models.OneToOneField(User, related_name='account', on_delete=models.CASCADE )
```
Comment (tallninja) [inline blog/models.py:6]: Why Is it a OneToOne relationship ? I thought a user can have many accounts.

**Review by tallninja — COMMENTED** [review]
Issues with the relationships. Please correct me if I'm wrong

**Pull request merged**